### 1) f001 — Total Characters

**What:** Counts every character in a text, including spaces and punctuation.

**Approach:**


In [4]:
text = "الذكاء الاصطناعي رائع!"
len(text)

22

**Result:** `22`

**Interpretation:** Longer texts → higher counts. Useful to normalize other metrics (e.g., ratios).

---

### 2) f004 — Whitespace / Characters Ratio

**What:** How much of the text is made of spaces.

**Approach:**

In [5]:
text = "الذكاء الاصطناعي رائع جدا"
whites = sum(ch.isspace() for ch in text)
ratio = whites / len(text)
ratio

0.12

**Result:** `0.12`

**Interpretation:** Higher ratio → shorter words or many small tokens; lower ratio → dense style.

---

### 3) f013 — Hapax Ratio

**What:** Fraction of words that appear exactly once.

**Approach:**


In [6]:
from collections import Counter
tokens = "الذكاء الاصطناعي الذكاء هو المستقبل".split()
freq = Counter(tokens)
hapax_ratio = sum(1 for w,c in freq.items() if c==1) / len(tokens)
hapax_ratio

0.6

**Result:** `0.6`

**Interpretation:** Higher ratio → richer vocabulary; lower → repetitive language.

---

### 4) f022 — Entropy of Word Frequencies

**What:** Shannon entropy = how evenly word frequencies are distributed.

**Approach:**

In [8]:
import math
p = [c/sum(freq.values()) for c in freq.values()]
H = -sum(pi*math.log2(pi) for pi in p)
H

1.9219280948873623

**Result:** `1.92`

**Interpretation:** Higher H → diverse word usage; lower H → few words dominate.

---

### 5) f025 — Single Quotes Count

**What:** Counts `'`, ‘`, or ’` characters.

**Approach:**

In [9]:
text = "قال: ‘مرحباً’"
text.count("'") + text.count("‘") + text.count("’")

2




**Result:** `2`

**Interpretation:** Used to detect quotation frequency or dialogue style.

---


### 6) f034 — Total Sentences

**What:** Number of sentence boundaries.

**Approach:**

In [10]:
import regex as re
sents = re.split(r"[.!؟]+", "هل تحب البرمجة؟ أنا أحبها. ممتاز!")
len([s for s in sents if s.strip()])

3

**Result:** `3`

**Interpretation:** Higher → complex text; lower → short statements.

---

### 7) f043 — Number of Nouns

**What:** How many tokens CAMeL marks as nouns.

**Approach:**

In [12]:
%pip install camel-tools

  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 3.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.7/43.7 kB 2.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 124.7/124.7 kB 10.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.2/3.2 MB 50.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.0/18.0 MB 90.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.4/9.4 MB 109.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 608.4/608.4 kB 39.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 122.3/122.3 kB 8.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 87.3 MB/s eta 0:00:00
  Created wheel for docopt: filename=docopt-0.6.2-py2.py3-none-any.whl size=13706 sha256=6fde67b2402310928f9d0c6d891da78ed9e34c7494b5167b25ef1ed171c9b1b4
  Stored in directory: /root/.cache/pip/wheels/1a/bf/a1/4cee4f7678c68c5875ca89eaccf460593539

In [1]:
!camel_data -i light

The following packages will be installed: 'morphology-db-lev-01', 'morphology-db-msa-s31', 'disambig-mle-calima-egy-r13', 'morphology-db-msa-r13', 'morphology-db-glf-01', 'disambig-mle-calima-msa-r13', 'dialectid-model26', 'morphology-db-egy-r13'
Extracting package 'morphology-db-lev-01': 100% 10.6M/10.6M [00:00<00:00, 276MB/s]
Extracting package 'morphology-db-msa-s31': 100% 44.8M/44.8M [00:00<00:00, 377MB/s]
Extracting package 'disambig-mle-calima-egy-r13': 100% 27.2M/27.2M [00:00<00:00, 427MB/s]
Extracting package 'morphology-db-msa-r13': 100% 40.5M/40.5M [00:00<00:00, 325MB/s]
Extracting package 'morphology-db-glf-01': 100% 7.98M/7.98M [00:00<00:00, 378MB/s]
Extracting package 'disambig-mle-calima-msa-r13': 100% 88.7M/88.7M [00:00<00:00, 344MB/s]
Extracting package 'dialectid-model26': 100% 371M/371M [00:05<00:00, 70.4MB/s]
Extracting package 'morphology-db-egy-r13': 100% 67.3M/67.3M [00:00<00:00, 346MB/s]


In [5]:
from camel_tools.disambig.mle import MLEDisambiguator
mle = MLEDisambiguator.pretrained()
analyses = mle.disambiguate(["الذكاء","الاصطناعي","يتطور"])
analyses[0]

DisambiguatedWord(word='الذكاء', analyses=[ScoredAnalysis(score=1.0, analysis={'diac': 'الذَكاء', 'lex': 'ذَكاء', 'bw': 'ال/DET+ذَكاء/NOUN', 'gloss': 'the+cleverness', 'pos': 'noun', 'prc3': '0', 'prc2': '0', 'prc1': '0', 'prc0': 'Al_det', 'per': 'na', 'asp': 'na', 'vox': 'na', 'mod': 'na', 'stt': 'd', 'cas': 'u', 'enc0': '0', 'rat': 'i', 'source': 'lex', 'form_gen': 'm', 'form_num': 's', 'd3seg': 'ال+_ذَّكاء', 'caphi': '2_a_dh_dh_a_k_aa_2', 'd1tok': 'الذَّكاء', 'd2tok': 'الذَّكاء', 'pos_logprob': -0.4344233, 'd3tok': 'ال+_ذَكاء', 'd2seg': 'الذَّكاء', 'pos_lex_logprob': -5.22446, 'num': 's', 'ud': 'NOUN', 'gen': 'm', 'catib6': 'NOM', 'root': 'ذ.ك.#', 'bwtok': 'ال+_ذَكاء', 'pattern': 'ال1َ2اء', 'lex_logprob': -5.22446, 'atbtok': 'الذَّكاء', 'atbseg': 'الذَّكاء', 'd1seg': 'الذَّكاء', 'stem': 'ذَكاء', 'stemgloss': 'cleverness', 'stemcat': 'N0_Nh'}, diac='الذَكاء', pos_lex_logprob=-5.22446, lex_logprob=-5.22446)])

In [7]:
analyses[1]

DisambiguatedWord(word='الاصطناعي', analyses=[ScoredAnalysis(score=1.0, analysis={'diac': 'الاِصْطِناعِيِّ', 'lex': 'ٱِصْطِناعِيّ', 'bw': 'ال/DET+ٱِصْطِناعِيّ/ADJ+ِ/CASE_DEF_GEN', 'gloss': 'the+artificial;synthetic+[def.gen.]', 'pos': 'adj', 'prc3': '0', 'prc2': '0', 'prc1': '0', 'prc0': 'Al_det', 'per': 'na', 'asp': 'na', 'vox': 'na', 'mod': 'na', 'stt': 'd', 'cas': 'g', 'enc0': '0', 'rat': 'n', 'source': 'lex', 'form_gen': 'm', 'form_num': 's', 'd3seg': 'ال+_اِصْطِناعِيِّ', 'caphi': '2_a_l_i_s._t._i_n_aa_3_i_y_y_i', 'd1tok': 'الاِصْطِناعِيِّ', 'd2tok': 'الاِصْطِناعِيِّ', 'pos_logprob': -0.9868824, 'd3tok': 'ال+_اِصْطِناعِيِّ', 'd2seg': 'الاِصْطِناعِيِّ', 'pos_lex_logprob': -4.446309, 'num': 's', 'ud': 'ADJ', 'gen': 'm', 'catib6': 'NOM', 'root': 'ص.ن.ع', 'bwtok': 'ال+_ٱِصْطِناعِيّ_+ِ', 'pattern': 'الٱِ1ْطِ2ا3ِيِّ', 'lex_logprob': -4.446309, 'atbtok': 'الاِصْطِناعِيِّ', 'atbseg': 'الاِصْطِناعِيِّ', 'd1seg': 'الاِصْطِناعِيِّ', 'stem': 'ٱِصْطِناعِيّ', 'stemgloss': 'artificial;synthetic',

In [9]:
analyses[2]

DisambiguatedWord(word='يتطور', analyses=[ScoredAnalysis(score=1.0, analysis={'diac': 'يَتَطَوَّر', 'lex': 'تَطَوَّر', 'bw': 'يَ/IV3MS+تَطَوَّر/IV', 'gloss': 'he;it+develop;advance', 'pos': 'verb', 'prc3': '0', 'prc2': '0', 'prc1': '0', 'prc0': '0', 'per': '3', 'asp': 'i', 'vox': 'a', 'mod': 'u', 'stt': 'na', 'cas': 'na', 'enc0': '0', 'rat': 'n', 'source': 'lex', 'form_gen': 'm', 'form_num': 's', 'd3seg': 'يَتَطَوَّر', 'caphi': 'y_a_t_a_t._a_w_w_a_r', 'd1tok': 'يَتَطَوَّر', 'd2tok': 'يَتَطَوَّر', 'pos_logprob': -1.023208, 'd3tok': 'يَتَطَوَّر', 'd2seg': 'يَتَطَوَّر', 'pos_lex_logprob': -4.6224, 'num': 's', 'ud': 'VERB', 'gen': 'm', 'catib6': 'VRB', 'root': 'ط.#.ر', 'bwtok': 'يَ+_تَطَوَّر', 'pattern': 'يَتَ1َوَّ3', 'lex_logprob': -4.6224, 'atbtok': 'يَتَطَوَّر', 'atbseg': 'يَتَطَوَّر', 'd1seg': 'يَتَطَوَّر', 'stem': 'تَطَوَّر', 'stemgloss': 'develop;advance', 'stemcat': 'IV'}, diac='يَتَطَوَّر', pos_lex_logprob=-4.6224, lex_logprob=-4.6224)])

In [6]:
sum(1 for d in analyses if d.analyses[0].analysis['pos']=='noun')

1

**Result:** `1`

**Interpretation:** Measures nominal density—texts with many nouns feel more factual or descriptive.

---

### 8) f046 — Number of Adverbs

**What:** Count tokens tagged `adv`.

**Approach:** (same pattern as above)

In [10]:
sum(1 for d in analyses if d.analyses[0].analysis['pos']=='adv')

0

**Result:** `0`

**Interpretation:** More adverbs → narrative or expressive tone.

---

### 9) f055 — Noun / Verb Ratio

**What:** Nouns ÷ Verbs.

**Approach:**

In [11]:
nouns = sum(1 for d in analyses if d.analyses[0].analysis['pos']=='noun')
verbs = sum(1 for d in analyses if d.analyses[0].analysis['pos']=='verb')
nouns/verbs if verbs else float("inf")

1.0

**Result:** `1`

**Interpretation:** High ratio → descriptive; low → action-heavy text.

---

### 10) f064 — Nominative Case Count

**What:** How many tokens are case-marked `nom`.

**Approach:** Use CAMeL `a.get('cas')`.

**Result:** depends on diacritics.

**Interpretation:** Higher nominative → many subjects or main-clause nouns.

---

### 11) f067 — Singular Words Count

**What:** Tokens marked as singular (`num='s'`).

**Approach:** `sum(1 for a in anal if a.get('num')=='s')`

**Result:** e.g. `5`

**Interpretation:** Plurality patterns → talking about individuals vs groups.

---

### 12) f076 — Passive Sentences

**What:** Count of sentences containing any passive verb (`vox='p'`).

**Approach:**

In [29]:
from camel_tools.disambig.mle import MLEDisambiguator
from camel_tools.tokenizers.word import simple_word_tokenize
import regex as re

mle = MLEDisambiguator.pretrained()

PASSIVE_HINT = re.compile(r'^[^\W\d_]\u064F[^\W\d_]\u0650[^\W\d_]\u064E$', re.UNICODE)

def has_passive_voice(tokens, disamb_words):
    # 1) morphology voice
    for w in disamb_words:
        for sc in getattr(w, 'analyses', []):
            if (sc.analysis.get('vox') or '').lower().startswith('p'):
                return True
    # 2) surface pattern fallback (works on fully diacritized Form I perfect)
    return any(PASSIVE_HINT.match(tok) for tok in tokens)

def count_passive_sentences(sentences):
    tokenized = [simple_word_tokenize(s) for s in sentences]
    disamb    = [mle.disambiguate(toks) for toks in tokenized]
    return sum(has_passive_voice(toks, dw) for toks, dw in zip(tokenized, disamb))

sentences = ["كُتِبَ الدرس.", "قرأ الطالب الدرس."]
count_passive_sentences(sentences)


1

**Result:** `1`

**Interpretation:** Higher → impersonal or formal writing style.

---

### 13) f085 — Sentence Length Variance

**What:** Variance of token counts per sentence.

**Approach:**

In [30]:
import numpy as np
sents = ["أنا أحب البرمجة", "الذكاء الاصطناعي يتطور بسرعة"]
lengths = [len(s.split()) for s in sents]
np.var(lengths)

0.25

**Result:** `0.25`

**Interpretation:** Low variance → uniform rhythm; high → mixed short/long sentences.

---

### ً14) f106 — Tanween Frequency

**What:** Counts Arabic tanween diacritics (ً ٌ ٍ).

**Approach:**

In [31]:
text = "كِتابٌ جميلٌ"
sum(ch in "ًٌٍ" for ch in text)

2

**Result:** `2`

**Interpretation:** Shows presence of full diacritization; high → formal/religious text.

---

### 15) f109 — Link Frequency

**What:** Number of URLs.

**Approach:**

In [32]:
import regex as re
text = "شاهد https://example.com و http://cnn.com"
len(re.findall(r"https?://\S+", text))

2


**Result:** `2`

**Interpretation:** High → informational or social-media style.

---

### 16) f088 — Semantic Similarity between Sentences

**What:** Average cosine similarity between consecutive sentence embeddings.

**Approach:**

In [35]:
%pip install -U "transformers==4.36.2" "sentence-transformers==2.5.1" torch

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 126.8/126.8 kB 8.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.2/8.2 MB 50.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 156.5/156.5 kB 11.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 899.7/899.7 MB 549.7 kB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 594.3/594.3 MB 3.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.2/10.2 MB 81.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 88.0/88.0 MB 9.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 954.8/954.8 kB 45.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 193.1/193.1 MB 6.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 56.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.6/63.6 MB 12.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 267.5/267.5 MB 4.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 28

In [1]:
from sentence_transformers import SentenceTransformer
import torch, numpy as np
model = SentenceTransformer("Omartificial-Intelligence-Space/Arabic-arabert-all-nli-triplet")
sents = ["الذكاء الاصطناعي يتطور بسرعة.", "هذا المجال مهم جداً."]
emb = model.encode(sents, convert_to_tensor=True, normalize_embeddings=True)
np.mean([float(torch.nn.functional.cosine_similarity(emb[i], emb[i+1], dim=0)) for i in range(len(emb)-1)])

/usr/local/lib/python3.12/dist-packages/transformers/utils/generic.py:441: FutureWarning: `torch.utils._pytree._register_pytree_node` is deprecated. Please use `torch.utils._pytree.register_pytree_node` instead.
  _torch_pytree._register_pytree_node(
/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/229 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/205 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

/usr/local/lib/python3.12/dist-packages/huggingface_hub/file_download.py:942: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


config.json:   0%|          | 0.00/637 [00:00<?, ?B/s]

/usr/local/lib/python3.12/dist-packages/transformers/utils/generic.py:309: FutureWarning: `torch.utils._pytree._register_pytree_node` is deprecated. Please use `torch.utils._pytree.register_pytree_node` instead.
  _torch_pytree._register_pytree_node(
/usr/local/lib/python3.12/dist-packages/transformers/utils/generic.py:309: FutureWarning: `torch.utils._pytree._register_pytree_node` is deprecated. Please use `torch.utils._pytree.register_pytree_node` instead.
  _torch_pytree._register_pytree_node(
/usr/local/lib/python3.12/dist-packages/huggingface_hub/file_download.py:942: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


model.safetensors:   0%|          | 0.00/541M [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/695 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/296 [00:00<?, ?B/s]

0.3026711046695709

**Result:** `0.31`

**Interpretation:** Near 1 → sentences are semantically similar (cohesive); near 0 → topic shifts.

---

### 17) f097 — BERT Token Embedding Similarity

**What:** Mean cosine similarity of consecutive subword embeddings within a sentence (AraBERT).

**Approach:**

In [4]:
from transformers import AutoTokenizer, AutoModel
import torch, numpy as np
tok = AutoTokenizer.from_pretrained("aubmindlab/bert-base-arabertv02")
mod = AutoModel.from_pretrained("aubmindlab/bert-base-arabertv02")
inputs = tok("كرة القدم كاس العالم", return_tensors="pt")
with torch.no_grad(): h = mod(**inputs).last_hidden_state.squeeze(0)
np.mean(torch.nn.functional.cosine_similarity(h[:-1], h[1:], dim=1).numpy())


0.54778737

**Result:** `0.55`

**Interpretation:** High → smooth token transitions; low → abrupt or code-switched text.